File: main_project/client/java/OceAdapter/src/main/java/com/xiaonei/talk/MucGateAdapter.java
package com.xiaonei.talk;

import java.util.ArrayList;
import java.util.List;
import java.util.Vector;

import com.xiaonei.talk.muc.MucGateManagerPrx;
import com.xiaonei.talk.muc.MucGateManagerPrxHelper;
import com.xiaonei.talk.muc.MucRoomId;
import com.xiaonei.talk.muc.Permisions;
import xce.util.channel.Channel;
import xce.util.service.ServiceAdapter;

public class MucGateAdapter extends ServiceAdapter {
	private static final String endpoints = "@MucGate";
	private Vector<MucGateManagerPrx> managersOneway = new Vector<MucGateManagerPrx>();;
	private Vector<MucGateManagerPrx> managersTwoway = new Vector<MucGateManagerPrx>();;

	private static MucGateAdapter instance = new MucGateAdapter();

	public static MucGateAdapter instance() {
		return instance;
	}

	MucGateAdapter() {
		super(endpoints, 10, Channel.newTalkChannel(endpoints));

	}

	public MucGateManagerPrx getManager(int id) {
		return locateProxy(managersTwoway, "M", id, Channel.Invoke.Twoway,
				MucGateManagerPrxHelper.class, 300);
	}
	
	public MucGateManagerPrx getManagerOneway(int id) {
		return locateProxy(managersOneway, "M", id, Channel.Invoke.Oneway,
				MucGateManagerPrxHelper.class, 300);
	}

	public Permisions Int2Permision(int p){
		if(1 == p){
			return Permisions.PMember;
		}
		else if(2 == p){
			return Permisions.PAdmin;
		}
		else if(3 == p){
			return Permisions.POwner;
		}
		else if(5 == p || 8 == p){
			return Permisions.PNone;
		}
		else{
			return Permisions.PMember;
		}
	}
	public void ChatSettingChange(int userid, int groupid, int set){
		MucRoomId r = new MucRoomId();
		r.domain = "group.talk.xiaonei.com";
		r.roomname = Integer.toString(groupid);
		getManagerOneway(groupid).ChatSetingChange(userid, r, set);
	}
	public void CreateGroup(int groupid) {
		MucRoomId r = new MucRoomId();
		r.domain = "group.talk.xiaonei.com";
		r.roomname = Integer.toString(groupid);
		getManagerOneway(groupid).CreateGroup(r);
	}
	public void CreateGroupForUgc(int groupowner, int groupid, int[] inviteusers, int invited_permision){
		MucRoomId r = new MucRoomId();
		r.domain = "group.talk.xiaonei.com";
		r.roomname = Integer.toString(groupid);
		Permisions p = Int2Permision(invited_permision);
		getManagerOneway(groupid).CreateGroupForUgc(groupowner, r, inviteusers, p);
	}
	public void RemoveGroup(int groupid){
		MucRoomId r = new MucRoomId();
		r.domain = "group.talk.xiaonei.com";
		r.roomname = Integer.toString(groupid);
		getManagerOneway(groupid).RemoveGroup(r);
	}
	
	public void ReloadGroupInfo(int groupid, String groupname){
		getManagerOneway(groupid).ReloadGroupInfo(groupid, groupname);
	}
	public void AddGroupMember(int userid, int groupid, int permision){
		MucRoomId r = new MucRoomId();
		r.domain = "group.talk.xiaonei.com";
		r.roomname = Integer.toString(groupid);
		Permisions p = Int2Permision(permision);
		getManagerOneway(groupid).AddGroupMember(userid, r, p);
	}
	public void AddGroupMemberBySeq(int[] userids, int groupid, int permision){
		if(userids.length == 0){
			return;
		}
		MucRoomId r = new MucRoomId();
		r.domain = "group.talk.xiaonei.com";
		r.roomname = Integer.toString(groupid);
		Permisions p = Int2Permision(permision);
		getManagerOneway(groupid).AddGroupMemberBySeq(userids, r, p);
	}
	public void RemoveGroupMember(int userid, int groupid){
		MucRoomId r = new MucRoomId();
		r.domain = "group.talk.xiaonei.com";
		r.roomname = Integer.toString(groupid);
		getManagerOneway(groupid).RemoveGroupMember(userid, r);
	}
	public void SetPermision(int operateuser, int groupid, int targetuser, int permision){
		MucRoomId r = new MucRoomId();
		r.domain = "group.talk.xiaonei.com";
		r.roomname = Integer.toString(groupid);
		Permisions p = Int2Permision(permision);
		getManagerOneway(groupid).SetPermision(operateuser, r, targetuser, p);
	}

	public List<Integer> GetUserListByGroupId(int groupid){
		List<Integer> result = new ArrayList<Integer>();
		int [] v = getManager(groupid).GetUserListByGroupId(groupid);
		for(int i : v){
			result.add(i);
		}
		return result;
	}

	public List<Integer> GetGroupListByUserId(int userid){
		List<Integer> result = new ArrayList<Integer>();
		int [] v = getManager(userid).GetGroupListByUserId(userid);
		for(int i : v){
			result.add(i);
		}
		return result;
	}

}

File: main_project/OceCxxAdapter/src/MucGateAdapter.cpp
#include "MucGateAdapter.h"
#include <boost/lexical_cast.hpp>
#include <string>
using namespace MyUtil;
using namespace com::xiaonei::talk::muc;
using namespace boost;
using namespace std;
MucGateManagerPrx MucGateAdapter::getManager(int id) {
  //  return locate<MucGateManagerPrx> (_managers, "M", id, TWO_WAY);
	return getProxy(id);
}

MucGateManagerPrx MucGateAdapter::getManagerOneway(int id) {
  //  return locate<MucGateManagerPrx> (_managersOneway, "M", id, ONE_WAY);
  return getProxyOneway(id);
}
MucRoomIdPtr MucGateAdapter::MakeRoomIdPtr(int groupid, const string& domain){
  string roomname = "";
  try{
		//change_name_request
    //roomname = boost::lexical_cast<string>(groupid) + "_" + groupname;
    roomname = boost::lexical_cast<string>(groupid);
  }
  catch (std::exception& e){
    MCE_WARN("MucGateAdapter::MageRoomIdPtr-->boost::lexical_cast to string exception  val = " << groupid);
    return NULL;
  }
  MucRoomIdPtr ans = new MucRoomId();
  ans->roomname = roomname;
  ans->domain = domain;
  return ans;
}

Permisions MucGateAdapter::Int2Permision(int p){
	if(1 == p){
		p = 2;
	}
	else if(2 == p){
		p = 3;
	}
	else if(3 == p){
		p = 4;
	}
	else if(5 == p){
		p = 0;
	}
	else{
		p = 0;
	}
	return Permisions(p);
}

TalkUserPtr MucGateAdapter::GetUserById(int userid){
	return getManager(0)->GetUserById(userid);
}
TalkUserMap MucGateAdapter::GetUserBySeq(const MyUtil::IntSeq& uids){
	return getManager(0)->GetUserBySeq(uids);
}
//for memcache test

void MucGateAdapter::ChatSetingChange(int userid, int groupid, int set){
	MucRoomIdPtr roomid = MakeRoomIdPtr(groupid, GROUPDOMAIN);
  getManagerOneway(userid)->ChatSetingChange(userid, roomid, set);
}
void MucGateAdapter::CreateGroupForUgc(int groupowner, int groupid, const IntSeq& inviteusers, int permision){
	MucRoomIdPtr roomid = MakeRoomIdPtr(groupid, GROUPDOMAIN);
	Permisions p = Int2Permision(permision);
  getManagerOneway(groupowner)->CreateGroupForUgc(groupowner, roomid, inviteusers, p);
}

void MucGateAdapter::CreateGroup(int groupid){
	MucRoomIdPtr roomid = MakeRoomIdPtr(groupid, GROUPDOMAIN);
  getManagerOneway(groupid)->CreateGroup(roomid);
}
void MucGateAdapter::RemoveGroup(int groupid){
	MucRoomIdPtr roomid = MakeRoomIdPtr(groupid, GROUPDOMAIN);
  getManagerOneway(groupid)->RemoveGroup(roomid);
}

void MucGateAdapter::ReloadGroupInfo(int groupid, const string& groupname){
	//MucRoomIdPtr roomid = MakeRoomIdPtr(groupid, GROUPDOMAIN);
	getManagerOneway(groupid)->ReloadGroupInfo(groupid, groupname);
}
void MucGateAdapter::AddGroupMember(int userid, int groupid, int permision){
	MucRoomIdPtr roomid = MakeRoomIdPtr(groupid, GROUPDOMAIN);
	Permisions p = Int2Permision(permision);
  getManagerOneway(userid)->AddGroupMember(userid, roomid, p);
}
void MucGateAdapter::AddGroupMemberBySeq(const MyUtil::IntSeq& userids, int groupid, int permision){
	MucRoomIdPtr roomid = MakeRoomIdPtr(groupid, GROUPDOMAIN);
	Permisions p = Int2Permision(permision);
  getManagerOneway((*(userids.begin())))->AddGroupMemberBySeq(userids, roomid, p);
}
void MucGateAdapter::RemoveGroupMember(int userid, int groupid){
	MucRoomIdPtr roomid = MakeRoomIdPtr(groupid, GROUPDOMAIN);
  getManagerOneway(userid)->RemoveGroupMember(userid, roomid);
}
void MucGateAdapter::SetPermision(int operateuser, int groupid, int targetuser, int permision){
	MucRoomIdPtr roomid = MakeRoomIdPtr(groupid, GROUPDOMAIN);
	Permisions p = Int2Permision(permision);
  getManagerOneway(operateuser)->SetPermision(operateuser, roomid, targetuser, p);
}
//FOR IM
void MucGateAdapter::EnterRoom(const MucActiveUserPtr& enteruser){
  getManagerOneway(enteruser->jid->userId)->EnterRoom(enteruser);
}
bool MucGateAdapter::SetPermisionForIM(int operateuser, const MucRoomIdPtr& roomid, int targetuser, Permisions permision){
  return getManager(operateuser)->SetPermisionForIM(operateuser, roomid, targetuser, permision);
}
MucActiveUserSeq MucGateAdapter::GetUserByPermision(const MucRoomIdPtr& roomid, Permisions permision){
  int irid = GetIdFromMucRoomId(roomid);
	return getManager(irid)->GetUserByPermision(roomid, permision);
}

void MucGateAdapter::UserInviteUser(const JidPtr& opjid, const MucRoomIdPtr& roomid, int targetuser){
	return getManagerOneway(opjid->userId)->UserInviteUser(opjid, roomid, targetuser);
}
void MucGateAdapter::SendPrivateMsg(const JidPtr& sender, const MucUserIdentityPtr& recidentity, const string& msg){
	return getManagerOneway(sender->userId)->SendPrivateMsg(sender, recidentity, msg);
}
//FOR PAGER
MucUserIdentitySeq MucGateAdapter::GetRoomUserList(const MucRoomIdPtr& roomid){
  int irid = GetIdFromMucRoomId(roomid);
  return getManager(irid)->GetRoomUserList(roomid);
}
string MucGateAdapter::GetGroupName(int groupid){
	return getManager(groupid)->GetGroupName(groupid);
}
//IM && PAGER(groups.renren.com)
void MucGateAdapter::UserOnline(const JidPtr& onlinejid){
  getManagerOneway(onlinejid->userId)->UserOnline(onlinejid);
}
void MucGateAdapter::UserOffline(const JidPtr& offlinejid){
  getManagerOneway(offlinejid->userId)->UserOffline(offlinejid);
}
MucRoomIdSeq MucGateAdapter::GetUserRooms(int userid, const string& domain){
	MucRoomIdSeq ans;
  ans = getManager(userid)->GetUserRooms(userid, domain);
	MCE_DEBUG("MucGateAdapter::GetUserRooms --> ans.size = " << ans.size());
	return ans;
}
void MucGateAdapter::AwayRoom(const JidPtr& opjid, const MucRoomIdPtr& roomid){
  getManagerOneway(opjid->userId)->AwayRoom(opjid, roomid);
}
void MucGateAdapter::SendTalkMsg(const JidPtr& opjid, const MucRoomIdPtr& roomid, const string& msg){
  getManagerOneway(opjid->userId)->SendTalkMsg(opjid, roomid, msg);
}
//------------------------RESPONSE---------
// 这个notify 的几个方法，一定要按照roomid散，不然，gate内部的cache会有问题。
void MucGateAdapter::NotifyUserEnterRoom(const MucActiveUserPtr& activeuser, const MucActiveUserSeq& presenceUsers, int mucstatus){
  int irid = GetIdFromMucRoomId(activeuser->identity->roomid);
  getManagerOneway(irid)->NotifyUserEnterRoom(activeuser, presenceUsers, mucstatus);
}
void MucGateAdapter::NotifyUserAwayRoom(const MucActiveUserPtr& awayuser, const JidSeq& presenceUsers, int mucstatus){
  int irid = GetIdFromMucRoomId(awayuser->identity->roomid);
  getManagerOneway(irid)->NotifyUserAwayRoom(awayuser, presenceUsers, mucstatus);
}
void MucGateAdapter::NotifyGroupChat(const MucActiveUserPtr& activeuser, const JidSeq& presenceUsers, const string& msg){
  int irid = GetIdFromMucRoomId(activeuser->identity->roomid);
  getManagerOneway(irid)->NotifyGroupChat(activeuser, presenceUsers, msg);
}
void MucGateAdapter::NotifySetPermision(const MucActiveUserPtr& opuser, const JidSeq& presenceUsers, const MucActiveUserSeq& targetusers, Permisions permision){
  int irid = GetIdFromMucRoomId(opuser->identity->roomid);
  getManagerOneway(irid)->NotifySetPermision(opuser, presenceUsers, targetusers, permision);
}
void MucGateAdapter::InviteIMusers(const MucRoomIdPtr& roomid, const JidSeq& imusers){
  int irid = GetIdFromMucRoomId(roomid);
  getManagerOneway(irid)->InviteIMUsers(roomid, imusers);
}

void MucGateAdapter::NotifyPrivateChat(const MucActiveUserPtr& sender, int targetuser, const string& msg){
  int irid = GetIdFromMucRoomId(sender->identity->roomid);
	getManagerOneway(irid)->NotifyPrivateChat(sender, targetuser, msg);
}
void MucGateAdapter::NotifyNicknameError(const MucRoomIdPtr& roomid, const JidPtr& jid){
  int irid = GetIdFromMucRoomId(roomid);
	getManagerOneway(irid)->NotifyNicknameError(roomid, jid);
}
void MucGateAdapter::NotifyChatSetingChange(const MucRoomIdPtr& roomid, const JidSeq& jids, int set){
  int irid = GetIdFromMucRoomId(roomid);
	getManagerOneway(irid)->NotifyChatSetingChange(roomid, jids, set);
}


IntSeq MucGateAdapter::GetUserListByGroupId(int groupid) {
	return getManager(groupid)->GetUserListByGroupId(groupid);
}

IntSeq MucGateAdapter::GetGroupListByUserId(int userid) {
  return getManager(userid)->GetGroupListByUserId(userid);
}
Str2StrMap MucGateAdapter::GetGroupNameList(int userid) {
  return getManager(userid)->GetGroupNameList(userid);
}
